x/gov/src/keeper.rs

```diff
@@ -314,9 +355,302 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, M: Module, BK: BankKeeper<SK, M>>
+
+            for delegation in self
+                .staking_keeper
+                .delegations_iter(ctx, &voter)
+                .collect::<Vec<_>>()
+            {
+                let delegation = delegation?;
+
+                if let Some(validator) = curr_validators.get_mut(delegation.validator()) {
+                    // from cosmos: https://github.com/cosmos/cosmos-sdk/blob/d3f09c222243bb3da3464969f0366330dcb977a8/x/gov/keeper/tally.go#L51
+                    // There is no need to handle the special case that validator address equal to voter address.
+                    // Because voter's voting power will tally again even if there will deduct voter's voting power from validator.
+
+                    validator.delegator_deduction += delegation.shares(); // TODO: handle overflow?
+
+                    // delegation shares * bonded / total shares
+                    let voting_power = delegation
+                        .shares()
+                        .mul(Decimal256::new(validator.bounded_tokens))
+                        .div(validator.delegator_shares);
+
+                    for VoteOptionWeighted { option, weight } in &vote_options {
+                        let result_option = tally_results.get_mut(option);
+
+                        *result_option += voting_power * Decimal256::from(weight.clone());
+                    }
+
+                    total_voting_power += voting_power;
+                }
+
+                vote_del(ctx, &self.store_key, proposal_id, &voter)?;
+            }
+        }
+
+        for (
+            _,
+            ValidatorGovInfo {
+                address: _,
+                bounded_tokens,
+                delegator_shares,
+                delegator_deduction,
+                vote,
+            },
+        ) in &curr_validators
+        {
+            if vote.is_empty() {
+                continue;
+            }
+
+            let voting_power = (delegator_shares - delegator_deduction)
+                * Decimal256::new(bounded_tokens.clone())
+                / delegator_shares;
+
+            for VoteOptionWeighted { option, weight } in vote {
+                let result = tally_results.get_mut(option);
+                *result += voting_power * Decimal256::from(weight.clone());
+            }
+
+            total_voting_power += voting_power;
+        }
+
+        let tally_params = self.gov_params_keeper.try_get(ctx)?.tally;
+
+        let total_bonded_tokens = self.staking_keeper.total_bonded_tokens(ctx)?;
+
+        // If there is no staked coins, the proposal fails
+        if total_bonded_tokens.amount.is_zero() {
+            return Ok((false, false, tally_results.result()));
+        }
+
+        // If there is not enough quorum of votes, the proposal fails
+        let percent_voting = total_voting_power / Decimal256::new(total_bonded_tokens.amount);
+        if percent_voting < tally_params.quorum {
+            return Ok((false, true, tally_results.result()));
```

Comment:
It seems like the method is a getter? I suggest that here we have a conversion. If yes, then it is better to change `to_result` or `to_tally_result`. It is more readable variants.

Reply:
Renamed to `to_result`